### PR TITLE
[CU-357bfp6] Get population statistic api

### PR DESCRIPTION
**Pull request by vitvara**
Implement Get population statistic endpoint `/population/statistic`
Screen Shot 2565-10-29 at 21 49 26

### CODE REVIEW

**Comment by sirateek** [comment]
Task linked: [CU-357bfp6 Implement "Get population statistics API."](https://app.clickup.com/t/357bfp6)

File: config/config.go [hunk 1]
```diff
@@ -11,11 +11,13 @@ import (
 type Config struct {
 	Env            string
 	GIN_MODE       string `envconfig:"GIN_MODE" default:"release"`
+	LISTENING_PORT string `envconfig:"LISTENING_PORT":default:"3000"`
```
Comment (sirateek) [inline config/config.go:14]: Invalid syntax

```suggestion
	LISTENING_PORT string `envconfig:"LISTENING_PORT" default:"3000"`
```

File: handler/middleware.go [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+	"github.com/sirupsen/logrus"
+)
+
+func ValidateAPIKey(apiKey string) gin.HandlerFunc {
+	return func(c *gin.Context) {
+		if c.GetHeader("Authorization") != apiKey {
+			errMessage := "APIKEY not found"
+			logrus.Warn(errMessage, ", APIKEY: ", c.GetHeader("Authorization"))
+			c.JSON(http.StatusUnauthorized, gin.H{"status": 401, "message": "Authentication failed"})
+			c.Abort()
```
Comment (sirateek) [inline handler/middleware.go:16]:
```suggestion
			c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"status": 401, "message": "Authentication failed"})
```

File: main.go [hunk 3]
```diff
@@ -1,7 +1,46 @@
...
+		return
+	}
+	// New Repository
+	populationRepository := repository.NewPopulationRepository(mysql)
+
+	populationService := service.NewPopulationService(populationRepository)
+
+	populationHandler := handler.NewPopulationHandler(populationService)
+
+	server := gin.Default()
```
Comment (sirateek) [inline main.go:39]: Please use httpserver from common-go
```suggestion
	server := httpserver.NewHttpServer()
```

File: config/config.go [hunk 4]
```diff
@@ -11,11 +11,13 @@ import (
...
 	MYSQL_HOSTNAME string `envconfig:"MYSQL_HOSTNAME"`
 	MYSQL_PORT     string `envconfig:"MYSQL_PORT"`
 	MYSQL_USERNAME string `envconfig:"MYSQL_USERNAME"`
 	MYSQL_PASSWORD string `envconfig:"MYSQL_PASSWORD"`
 	MYSQL_DATABASE string `envconfig:"MYSQL_DATABASE"`
+	API_KEY        string `envconfig:"API_KEY`
```
Comment (sirateek) [inline config/config.go:20]:
```suggestion
	API_KEY        string `envconfig:"API_KEY"`
```

File: repository/population.go [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+	if err != nil {
+		p.errorMessage(err, "QueryPeopleCommitedTheVote")
+		return res, err
+	}
+	p.queryLog(fmt.Sprintf("Get total people commited the vote districtId: %s", districtId), "QueryPeopleCommitedTheVote")
+	return res, nil
+}
+
+func (p *populationRepository) QueryPeopleRightToVote(districtId string) (ResultQueryPeopleWithRightToVote, error) {
+	res := ResultQueryPeopleWithRightToVote{}
```
Comment (sirateek) [inline repository/population.go:107]: Instead of returning a struct. Can we return `int` instead?

File: repository/population.go [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+	if err != nil {
+		p.errorMessage(err, "QueryTotalPopulation")
+		return res, err
+	}
+	p.queryLog(fmt.Sprintf("Get total population districtId: %s", districtId), "QueryTotalPopulation")
+	return res, nil
+
+}
+
+func (p *populationRepository) QueryPeopleCommitedTheVote(districtId string) (ResultQueryPeopleCommitTheVote, error) {
```
Comment (sirateek) [inline repository/population.go:84]: Instead of returning a struct. Can we return `int` instead?

File: repository/population.go [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+	FROM District`
+	err := p.mysql.Select(&res, q)
+	if err != nil {
+		p.errorMessage(err, "QueryAllDistrict")
+		return res, err
+	}
+	p.queryLog("Select all district", "QueryAllDistrict")
+	return res, nil
+}
+func (p *populationRepository) QueryTotalPopulation(districtId string) (ResultQueryTotal, error) {
```
Comment (sirateek) [inline repository/population.go:68]: Instead of returning a struct. Can we return `int` instead?

File: repository/population.go [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+
+type ResultQueryTotal struct {
+	Total int64 `db:"Total"`
+}
+type ResultQueryPeopleWithRightToVote struct {
+	PeopleWithRightToVote int64 `db:"HaveRight"`
+}
+type ResultQueryPeopleCommitTheVote struct {
+	PeopleCommitTheVote int64 `db:"Commits"`
+}
```
Comment (sirateek) [inline repository/population.go:40]: This struct seems to be too specific. Please see the comment in function where you use it below krub.

File: service/population.go [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+			logrus.Error(err)
+		}
+		haveRight, err := p.repository.QueryPeopleRightToVote(districtId)
+		if err != nil {
+			logrus.Error(err)
+		}
+		commit, err := p.repository.QueryPeopleCommitedTheVote(districtId)
+		if err != nil {
+			logrus.Error(err)
+		}
```
Comment (sirateek) [inline service/population.go:43]: Should we skip to insert this result in array if some error happen during these call?

File: repository/population.go [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+			AND p.DistrictId=?
+			AND p.Nationality="Thai"
+		) AS c
+	`
+	err := p.mysql.Get(&res, q, districtId)
+	if err != nil {
+		p.errorMessage(err, "QueryPeopleRightToVote")
+		return res, err
+	}
+	p.queryLog(fmt.Sprintf("Get total people have right to vote districtId: %s", districtId), "QueryPeopleRightToVote")
```
Comment (abounk) [inline repository/population.go:113]: Will it better if we make the logging to start with a function name? ex. ("QueryPeopleRightToVote: Get total people have right to vote districtId: 1")

Comment (vitvara) [reply repository/population.go:113]: Screen Shot 2565-11-01 at 19 24 29
It also has a function-name, and module name displayed on the back of the log message. Is this ok for you?

Comment (abounk) [reply repository/population.go:113]: Oh, this should work!

**Review by abounk — APPROVED** [review]
hod mak2

File: repository/population.go [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+func NewPopulationRepository(mysql *sqlx.DB) PopulationRepository {
+	return &populationRepository{
+		mysql: mysql,
+	}
+}
+
+type ResultQueryDistrictIdAndName struct {
+	LocationID string `db:"DistrictID"`
+	Location   string `db:"Name"`
+}
```
Comment (sirateek) [inline repository/population.go:30]: Move this to model

Comment (sirateek) [reply repository/population.go:30]: You can also create a full model if the data and then partially use it.

File: service/population.go (hunk 9 above)
Comment (sirateek) [reply service/population.go:43]: What about this?

File: handler/middleware.go [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+	"github.com/gin-gonic/gin"
+	"github.com/sirupsen/logrus"
+)
+
+func ValidateAPIKey(apiKey string) gin.HandlerFunc {
+	return func(c *gin.Context) {
+		if c.GetHeader("Authorization") != apiKey {
+			errMessage := "APIKEY not found"
+			logrus.Warn(errMessage, ", APIKEY: ", c.GetHeader("Authorization"))
+			c.JSON(http.StatusUnauthorized, gin.H{"status": 401, "message": "Authentication failed"})
```
[the remark this answers is not in the document]
Comment (sirateek) [reply handler/middleware.go]: Fixed

File: handler/middleware.go [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+			c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"status": 401, "message": "Authentication failed"})
```
[the remark this answers is not in the document]
Comment (sirateek) [reply handler/middleware.go]: Fixed

File: service/population.go (hunk 9 above)
Comment (sirateek) [reply service/population.go]: Fixed

**Review by sirateek — APPROVED** [review]

**Pull request merged**